cpp/src/parquet/encoding_test.cc

```diff
@@ -1276,5 +1282,128 @@ TEST(ByteStreamSplitEncodeDecode, InvalidDataTypes) {
+    auto decoder = MakeTypedDecoder<Type>(Encoding::DELTA_BINARY_PACKED, descr_.get());
+
+    encoder->Put(draws_, num_values_);
+    encode_buffer_ = encoder->FlushValues();
+
+    decoder->SetData(num_values_, encode_buffer_->data(),
+                     static_cast<int>(encode_buffer_->size()));
+    int values_decoded = decoder->Decode(decode_buf_, num_values_);
+    ASSERT_EQ(num_values_, values_decoded);
+    ASSERT_NO_FATAL_FAILURE(VerifyResults<c_type>(decode_buf_, draws_, num_values_));
+  }
+
+  void CheckRoundtripSpaced(const uint8_t* valid_bits,
+                            int64_t valid_bits_offset) override {
+    auto encoder =
+        MakeTypedEncoder<Type>(Encoding::DELTA_BINARY_PACKED, false, descr_.get());
+    auto decoder = MakeTypedDecoder<Type>(Encoding::DELTA_BINARY_PACKED, descr_.get());
+    int null_count = 0;
+    for (auto i = 0; i < num_values_; i++) {
+      if (!bit_util::GetBit(valid_bits, valid_bits_offset + i)) {
+        null_count++;
+      }
+    }
+
+    encoder->PutSpaced(draws_, num_values_, valid_bits, valid_bits_offset);
+    encode_buffer_ = encoder->FlushValues();
+    decoder->SetData(num_values_ - null_count, encode_buffer_->data(),
+                     static_cast<int>(encode_buffer_->size()));
+    auto values_decoded = decoder->DecodeSpaced(decode_buf_, num_values_, null_count,
+                                                valid_bits, valid_bits_offset);
+    ASSERT_EQ(num_values_, values_decoded);
+    ASSERT_NO_FATAL_FAILURE(VerifyResultsSpaced<c_type>(decode_buf_, draws_, num_values_,
+                                                        valid_bits, valid_bits_offset));
+  }
+
+ protected:
+  USING_BASE_MEMBERS();
+  std::vector<uint8_t> input_bytes_;
+  std::vector<uint8_t> output_bytes_;
+};
+
+using TestDeltaBitPackEncodingTypes = ::testing::Types<Int32Type, Int64Type>;
+TYPED_TEST_SUITE(TestDeltaBitPackEncoding, TestDeltaBitPackEncodingTypes);
+
+TYPED_TEST(TestDeltaBitPackEncoding, BasicRoundTrip) {
+  using T = typename TypeParam::c_type;
+  int values_per_block = 128;
+  int values_per_mini_block = 32;
+
+  // Size a multiple of miniblock size
+  ASSERT_NO_FATAL_FAILURE(this->Execute(values_per_mini_block * 10, 10));
+  // Size a multiple of block size
+  ASSERT_NO_FATAL_FAILURE(this->Execute(values_per_block * 10, 10));
+  // Size multiple of neither miniblock nor block size
+  ASSERT_NO_FATAL_FAILURE(
+      this->Execute((values_per_mini_block * values_per_block) + 1, 10));
+  ASSERT_NO_FATAL_FAILURE(this->Execute(0, 0));
+  ASSERT_NO_FATAL_FAILURE(this->ExecuteSpaced(
+      /*nvalues*/ 1234, /*repeats*/ 1, /*valid_bits_offset*/ 64,
+      /*null_probability*/ 0.1));
+
+  ASSERT_NO_FATAL_FAILURE(this->ExecuteBound(2000, 2000, 0));
+  ASSERT_NO_FATAL_FAILURE(this->ExecuteSpacedBound(
+      /*nvalues*/ 1234, /*repeats*/ 1, /*valid_bits_offset*/ 64,
+      /*null_probability*/ 0.1,
+      /*half_range*/ 0));
+
+  const uint32_t half_bitwidth = sizeof(T) * 4;
```

Comment:
Why only half the bitwidth? Ideally we would go up to the full bitwidth (`sizeof(T) * 8`  inclusive).

Reply:
Data in `ExecuteBound` is generated by:
```cpp
GenerateBoundData<c_type>(nvalues, draws_, -half_range, half_range, &data_buffer_);
```
So half bitwidth should cover the whole. Maybe I'm missing some edge case?

Reply:
Halving the bitwidth doesn't halve the range...

Reply:
Ugh, yeah.

Reply:
Changed to:
```cpp
  const uint32_t max_bitwidth = sizeof(T) * 8;
  for (uint32_t bitwidth = 0; bitwidth <= max_bitwidth; bitwidth += 4) {
    T half_range = (2 << (bitwidth - 1)) / 2;
   [...]
```

Reply:
This was causing:
```bash
/arrow/cpp/src/parquet/encoding_test.cc:1398:23: runtime error: shift exponent 4294967295 is too large for 32-bit type 'int'
```
So I changed it to:
```cpp
  using UT = std::make_unsigned_t<T>;
[...]
  const uint32_t max_bitwidth = sizeof(T) * 8;
  for (uint32_t bitwidth = 4; bitwidth <= max_bitwidth; bitwidth += 4) {
    UT half_range = (UT(1) << (bitwidth - 1)) / 2;
[...]
```